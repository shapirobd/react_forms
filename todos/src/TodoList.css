.TodoList {
	width: 75%;
	margin: 10px auto;
	margin-top: 0;
}
