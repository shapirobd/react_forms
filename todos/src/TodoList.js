import React, { useState } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import "./TodoList.css";
import uuid from "uuid/v4";

import { makeStyles } from "@material-ui/core/styles";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableContainer from "@material-ui/core/TableContainer";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";
import Paper from "@material-ui/core/Paper";

const TodoList = () => {
	const useStyles = makeStyles({
		table: {
			backgroundColor: "#BBB8B6",
		},
	});

	const classes = useStyles();
	const INITIAL_STATE = [];
	const [todos, setTodos] = useState(INITIAL_STATE);
	const addTodo = (todo) => {
		const { task } = todo;
		setTodos((todos) => [...todos, { task, id: uuid(), completed: false }]);
	};
	const removeTodo = (todo) => {
		setTodos(todos.filter((t) => t.id !== todo.id));
	};
	const editTodo = (oldTodo, newTodo) => {
		setTodos(
			todos.map((todo) => {
				if (todo.task === oldTodo) {
					return { task: newTodo.task };
				}
			})
		);
	};
	const completeTodo = (todo) => {
		setTodos(
			todos.map((t) => {
				if (todo.id === t.id) {
					return { ...todo, completed: true };
				}
			})
		);
	};
	console.log(todos);
	return (
		<div className="TodoList">
			<NewTodoForm addTodo={addTodo} />
			<TableContainer component={Paper}>
				<Table className={classes.table}>
					<TableBody>
						{todos.map((todo) => (
							<Todo
								task={todo.task}
								removeTodo={() => {
									removeTodo(todo);
								}}
								editTodo={editTodo}
								completeTodo={() => completeTodo(todo)}
								completed={todo.completed}
								key={todo.id}
							/>
						))}
					</TableBody>
				</Table>
			</TableContainer>
		</div>
	);
};

export default TodoList;
