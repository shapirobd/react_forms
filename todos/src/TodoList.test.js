import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

it("renders without crashing", () => {
	render(<TodoList />);
});

it("matches snapshot", () => {
	const { asFragment } = render(<TodoList />);
	expect(asFragment()).toMatchSnapshot();
});

it("should add new todo to list when form submits", () => {
	const { getByLabelText, queryByText } = render(<TodoList />);
	const input = getByLabelText("Add Todo");
	const btn = queryByText("Submit");
	expect(queryByText("Clean bedroom")).not.toBeInTheDocument();

	fireEvent.change(input, { target: { value: "Clean bedroom" } });
	fireEvent.click(btn);
	expect(queryByText("Clean bedroom")).toBeInTheDocument();
});

it("should remove a todo when click on X button", () => {
	const { queryByText, getByLabelText } = render(<TodoList />);
	const input = getByLabelText("Add Todo");
	const submitBtn = queryByText("Submit");
	fireEvent.change(input, { target: { value: "Make the bed" } });
	fireEvent.click(submitBtn);
	expect(queryByText("Make the bed")).toBeInTheDocument();

	const deleteBtn = queryByText("X");
	fireEvent.click(deleteBtn);
	expect(queryByText("Make the bed")).not.toBeInTheDocument();
});
