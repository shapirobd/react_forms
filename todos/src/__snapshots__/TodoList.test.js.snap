// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <div
      class="NewTodoForm"
    >
      <h3>
        Todo List
      </h3>
      <form>
        <label
          for="task"
        >
          Add Todo
        </label>
        <input
          id="task"
          name="task"
          placeholder="Enter new todo"
          type="text"
          value=""
        />
        <button>
          Submit
        </button>
      </form>
    </div>
    <div
      class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <table
        class="MuiTable-root makeStyles-table-2"
      >
        <tbody
          class="MuiTableBody-root"
        />
      </table>
    </div>
  </div>
</DocumentFragment>
`;
